src/snowflake/snowpark/_internal/data_source/datasource_partitioner.py

```diff
@@ -0,0 +1,246 @@
+            processed_upper_bound - processed_lower_bound
+        ) < 0:
+            actual_num_partitions = num_partitions
+        else:
+            actual_num_partitions = processed_upper_bound - processed_lower_bound
+            logger.warning(
+                "The number of partitions is reduced because the specified number of partitions is less than the difference between upper bound and lower bound."
+            )
+
+        # decide stride length
+        upper_stride = (
+            processed_upper_bound / decimal.Decimal(actual_num_partitions)
+        ).quantize(decimal.Decimal("1e-18"), rounding=ROUND_HALF_EVEN)
+        lower_stride = (
+            processed_lower_bound / decimal.Decimal(actual_num_partitions)
+        ).quantize(decimal.Decimal("1e-18"), rounding=ROUND_HALF_EVEN)
+        precise_stride = upper_stride - lower_stride
+        stride = int(precise_stride)
+
+        lost_num_of_strides = (
+            (precise_stride - decimal.Decimal(stride))
+            * decimal.Decimal(actual_num_partitions)
+            / decimal.Decimal(stride)
+        )
+        lower_bound_with_stride_alignment = processed_lower_bound + int(
+            (lost_num_of_strides / 2 * decimal.Decimal(stride)).quantize(
+                decimal.Decimal("1"), rounding=ROUND_HALF_UP
+            )
+        )
+
+        current_value = lower_bound_with_stride_alignment
+
+        partition_queries = []
+        for i in range(actual_num_partitions):
+            l_bound = (
+                f"{column} >= '{to_external_value(current_value, column_type)}'"
+                if i != 0
+                else ""
+            )
+            current_value += stride
+            u_bound = (
+                f"{column} < '{to_external_value(current_value, column_type)}'"
+                if i != actual_num_partitions - 1
+                else ""
+            )
+
+            if u_bound == "":
+                where_clause = l_bound
+            elif l_bound == "":
+                where_clause = f"{u_bound} OR {column} is null"
+            else:
+                where_clause = f"{l_bound} AND {u_bound}"
+
+            partition_queries.append(select_query + f" WHERE {where_clause}")
+
+        return partition_queries
+
+    @staticmethod
+    def generate_partition_with_predicates(
+        select_query: str, predicates: List[str]
+    ) -> List[str]:
+        return [select_query + f" WHERE {predicate}" for predicate in predicates]
+
+
+def to_internal_value(value: Union[int, str, float], column_type: DataType):
+    if isinstance(column_type, _NumericType):
+        return int(value)
+    elif isinstance(column_type, (TimestampType, DateType)):
+        # TODO: SNOW-1909315: support timezone
+        dt = parser.parse(value)
+        return int(dt.replace(tzinfo=pytz.UTC).timestamp())
+    else:
+        raise TypeError(f"unsupported column type for partition: {column_type}")
```

Comment:
we raised ValueError in the main branch

Reply:
i actually found out that we have a negative test for this, but this error was caught early before hit this line, I am going to remove the error raise here.
FYI, it was handled at line 130 before generate_partition() is called.